File: backend/utils/exchange_rates.py
"""
Real-time Exchange Rate Service
Uses free exchange rate API with 1-hour caching
"""

import aiohttp
from datetime import datetime, timedelta
from typing import Dict, Optional
import os


# Cache for exchange rates
_rate_cache: Dict = {
    "rates": {},
    "base": "USD",
    "last_updated": None
}

# Cache duration in hours
CACHE_DURATION_HOURS = 1


async def get_exchange_rates(base_currency: str = "USD") -> Dict[str, float]:
    """
    Fetch real-time exchange rates from free API
    Results are cached for 1 hour to minimize API calls
    
    Args:
        base_currency: Base currency code (default: USD)
        
    Returns:
        Dictionary of currency codes to exchange rates
    """
    global _rate_cache
    
    # Check if cache is valid
    if (
        _rate_cache["last_updated"] 
        and _rate_cache["base"] == base_currency
        and datetime.now() - _rate_cache["last_updated"] < timedelta(hours=CACHE_DURATION_HOURS)
    ):
        print(f"📊 Using cached exchange rates for {base_currency}")
        return _rate_cache["rates"]
    
    try:
        # Free API options (no API key required):
        # 1. exchangerate-api.com (free tier)
        # 2. open.er-api.com (free tier)
        
        async with aiohttp.ClientSession() as session:
            # Using open.er-api.com (no API key needed)
            url = f"https://open.er-api.com/v6/latest/{base_currency}"
            
            async with session.get(url, timeout=aiohttp.ClientTimeout(total=10)) as response:
                if response.status == 200:
                    data = await response.json()
                    
                    if data.get("result") == "success":
                        rates = data.get("rates", {})
                        
                        # Update cache
                        _rate_cache = {
                            "rates": rates,
                            "base": base_currency,
                            "last_updated": datetime.now()
                        }
                        
                        print(f"✅ Fetched fresh exchange rates for {base_currency}")
                        return rates
                    else:
                        print(f"⚠️ API returned error: {data.get('error-type')}")
                else:
                    print(f"⚠️ API returned status {response.status}")
                    
    except Exception as e:
        print(f"❌ Failed to fetch exchange rates: {e}")
    
    # Return fallback rates if API fails
    return get_fallback_rates(base_currency)


def get_fallback_rates(base: str = "USD") -> Dict[str, float]:
    """
    Fallback exchange rates when API is unavailable
    These are approximate rates and should be updated periodically
    """
    # Approximate rates as of Dec 2024
    usd_rates = {
        "USD": 1.0,
        "EUR": 0.92,
        "GBP": 0.79,
        "INR": 84.50,
        "JPY": 154.0,
        "AUD": 1.54,
        "CAD": 1.42,
        "CHF": 0.88,
        "CNY": 7.24,
        "SGD": 1.35,
        "AED": 3.67,
        "SAR": 3.75,
    }
    
    if base == "USD":
        return usd_rates
    
    # Convert to requested base
    if base in usd_rates:
        base_rate = usd_rates[base]
        return {
            currency: rate / base_rate 
            for currency, rate in usd_rates.items()
        }
    
    return usd_rates


async def convert_currency(
    amount: float, 
    from_currency: str, 
    to_currency: str
) -> float:
    """
    Convert amount from one currency to another
    
    Args:
        amount: Amount to convert
        from_currency: Source currency code
        to_currency: Target currency code
        
    Returns:
        Converted amount
    """
    if from_currency == to_currency:
        return amount
    
    rates = await get_exchange_rates(from_currency)
    rate = rates.get(to_currency, 1.0)
    
    return round(amount * rate, 2)


def get_cache_status() -> Dict:
    """Get current cache status for debugging"""
    return {
        "cached": _rate_cache["last_updated"] is not None,
        "base": _rate_cache["base"],
        "last_updated": _rate_cache["last_updated"].isoformat() if _rate_cache["last_updated"] else None,
        "rates_count": len(_rate_cache["rates"]),
        "cache_duration_hours": CACHE_DURATION_HOURS
    }


# Popular currency pairs for display
POPULAR_CURRENCIES = ["USD", "EUR", "GBP", "INR", "JPY", "AUD", "CAD", "CHF", "CNY", "SGD", "AED"]

File: frontend/src/utils/exchangeRates.ts
/**
 * Real-Time Exchange Rate Service
 * 
 * Uses Frankfurter API (European Central Bank rates)
 * - No API key required
 * - Updates daily (official ECB rates)
 * - Free and reliable
 * 
 * API: https://api.frankfurter.app/latest
 */

// Types are defined in ./currency.ts - this module provides real-time rate fetching

// Cached rates with timestamp
let cachedRates: Record<string, number> | null = null;
let lastFetchTime: number = 0;
const CACHE_DURATION = 60 * 60 * 1000; // 1 hour cache

// Fallback static rates (Dec 2024 approximate values)
// These are used when API is unavailable
const FALLBACK_RATES: Record<string, number> = {
    USD: 1.0,
    EUR: 0.95,    // 1 USD = 0.95 EUR
    GBP: 0.79,    // 1 USD = 0.79 GBP
    INR: 84.0,    // 1 USD = 84 INR
    JPY: 149.5,   // 1 USD = 149.5 JPY
    AUD: 1.54,    // 1 USD = 1.54 AUD
    CAD: 1.35,    // 1 USD = 1.35 CAD
    CHF: 0.88,    // 1 USD = 0.88 CHF
    CNY: 7.15,    // 1 USD = 7.15 CNY
    SGD: 1.35,    // 1 USD = 1.35 SGD
};

/**
 * Fetch latest exchange rates from API
 * Returns rates relative to USD
 */
export async function fetchExchangeRates(): Promise<Record<string, number>> {
    const now = Date.now();

    // Return cached rates if still valid
    if (cachedRates && (now - lastFetchTime) < CACHE_DURATION) {
        return cachedRates;
    }

    try {
        // Frankfurter API - ECB rates, base USD
        const response = await fetch('https://api.frankfurter.app/latest?from=USD');

        if (!response.ok) {
            throw new Error(`API error: ${response.status}`);
        }

        const data = await response.json();

        // Convert to our format: how many units of X = 1 USD
        const rates: Record<string, number> = { USD: 1.0 };

        if (data.rates) {
            for (const [currency, rate] of Object.entries(data.rates)) {
                rates[currency] = rate as number;
            }
        }

        // Cache the rates
        cachedRates = rates;
        lastFetchTime = now;

        console.log('✅ Exchange rates updated from ECB:', new Date().toLocaleString());

        return rates;
    } catch (error) {
        console.warn('⚠️ Failed to fetch exchange rates, using fallback:', error);
        return FALLBACK_RATES;
    }
}

/**
 * Get current exchange rates (cached or fallback)
 */
export function getExchangeRates(): Record<string, number> {
    return cachedRates || FALLBACK_RATES;
}

/**
 * Convert currency using real-time rates
 */
export async function convertCurrencyRealtime(
    amount: number,
    fromCurrency: string,
    toCurrency: string
): Promise<number> {
    if (fromCurrency === toCurrency) return amount;

    const rates = await fetchExchangeRates();

    // Both rates are relative to USD
    const fromRate = rates[fromCurrency] || 1;
    const toRate = rates[toCurrency] || 1;

    // Convert: source -> USD -> target
    const usdAmount = amount / fromRate;  // Convert to USD
    const targetAmount = usdAmount * toRate;  // Convert from USD to target

    return targetAmount;
}

/**
 * Sync version using cached rates
 */
export function convertCurrencySync(
    amount: number,
    fromCurrency: string,
    toCurrency: string
): number {
    if (fromCurrency === toCurrency) return amount;

    const rates = getExchangeRates();

    const fromRate = rates[fromCurrency] || 1;
    const toRate = rates[toCurrency] || 1;

    const usdAmount = amount / fromRate;
    const targetAmount = usdAmount * toRate;

    return targetAmount;
}

/**
 * Initialize rates on app load
 */
export function initializeExchangeRates(): void {
    fetchExchangeRates().then(() => {
        console.log('💱 Exchange rates initialized');
    }).catch(err => {
        console.warn('Failed to initialize rates:', err);
    });
}

/**
 * Get the current rate for a specific currency pair
 */
export async function getRate(fromCurrency: string, toCurrency: string): Promise<number> {
    const rates = await fetchExchangeRates();

    if (fromCurrency === 'USD') {
        return rates[toCurrency] || 1;
    }

    if (toCurrency === 'USD') {
        return 1 / (rates[fromCurrency] || 1);
    }

    // Cross rate: FROM -> USD -> TO
    const fromToUsd = 1 / (rates[fromCurrency] || 1);
    const usdToTarget = rates[toCurrency] || 1;

    return fromToUsd * usdToTarget;
}

/**
 * Format rate for display
 */
export function formatRate(rate: number): string {
    if (rate >= 100) {
        return rate.toFixed(2);
    } else if (rate >= 1) {
        return rate.toFixed(4);
    } else {
        return rate.toFixed(6);
    }
}
